common/datavalues2/src/data_value.rs

```diff
@@ -0,0 +1,318 @@
+            DataValue::Struct(_) => ValueType::Struct,
+        }
+    }
+    // convert to minialized data type
+    pub fn data_type(&self) -> DataTypePtr {
+        match self {
+            DataValue::Null => Arc::new(NullType {}),
+            DataValue::Boolean(_) => BooleanType::arc(),
+            DataValue::Int64(n) => {
+                if *n >= i8::MIN as i64 && *n <= i8::MAX as i64 {
+                    return Int8Type::arc();
+                }
+                if *n >= i16::MIN as i64 && *n <= i16::MAX as i64 {
+                    return Int16Type::arc();
+                }
+                if *n >= i32::MIN as i64 && *n <= i32::MAX as i64 {
+                    return Int32Type::arc();
+                }
+                Int64Type::arc()
+            }
+            DataValue::UInt64(n) => {
+                if *n <= u8::MIN as u64 {
+                    return UInt8Type::arc();
+                }
+                if *n <= u16::MIN as u64 {
+                    return UInt16Type::arc();
+                }
+                if *n <= u32::MIN as u64 {
+                    return UInt32Type::arc();
+                }
+                UInt64Type::arc()
+            }
+            DataValue::Float64(_) => Float64Type::arc(),
+            DataValue::String(_) => StringType::arc(),
+            DataValue::Array(x) => {
+                let inner_type = if x.is_empty() {
+                    UInt8Type::arc()
+                } else {
+                    x[0].data_type()
+                };
+                Arc::new(ArrayType::create(inner_type))
+            }
+            DataValue::Struct(x) => {
+                let names = (0..x.len()).map(|i| format!("{}", i)).collect::<Vec<_>>();
+                let types = x.iter().map(|v| v.data_type()).collect::<Vec<_>>();
+                Arc::new(StructType::create(names, types))
+            }
+        }
+    }
+
+    #[inline]
+    pub fn is_integer(&self) -> bool {
+        matches!(self, |DataValue::Int64(_)| DataValue::UInt64(_))
+    }
+
+    #[inline]
+    pub fn is_signed_integer(&self) -> bool {
+        matches!(self, DataValue::Int64(_))
+    }
+
+    #[inline]
+    pub fn is_unsigned_integer(&self) -> bool {
+        matches!(
+            self,
+                | DataValue::UInt64(_)
+        )
+    }
+
+    pub fn to_values(&self, size: usize) -> Result<Vec<DataValue>> {
+        Ok((0..size).map(|_| self.clone()).collect())
+    }
+
+    pub fn as_u64(&self) -> Result<u64> {
+        match self {
+            DataValue::Int64(v) if *v >= 0 => Ok(*v as u64),
+            DataValue::UInt64(v) => Ok(*v),
+            other => Result::Err(ErrorCode::BadDataValueType(format!(
+                "Unexpected type:{:?} to get u64 number",
+                other.value_type()
+            ))),
+        }
+    }
+
+    pub fn as_i64(&self) -> Result<i64> {
+        match self {
+            DataValue::Int64(v) => Ok(*v),
+            DataValue::UInt64(v) => Ok(*v as i64),
```

Comment:
Will we have overflow here?

Reply:
Yes, it will. `u64` as `i64` is allowed in rust, so we allow DataValue::u64 to DataValue::i64.